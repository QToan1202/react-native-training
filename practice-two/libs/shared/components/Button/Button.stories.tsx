import { ComponentMeta, ComponentStory } from '@storybook/react-native'
import { action } from '@storybook/addon-actions'

import { Button } from '../index'

export default {
  title: 'components/Button',
  component: Button,
} as ComponentMeta<typeof Button>

const Template: ComponentStory<typeof Button> = (args) => <Button {...args} />

export const Default = Template.bind({})
Default.args = {
  title: 'My Button',
  onPress: action('press'),
}

export const SecondaryButton = Template.bind({})
SecondaryButton.args = {
  ...Default.args,
  variant: 'secondary',
}

export const OutlineButton = Template.bind({})
OutlineButton.args = {
  ...Default.args,
  variant: 'tertiary',
}

export const FullButtonWithoutRadius = Template.bind({})
FullButtonWithoutRadius.args = {
  ...Default.args,
  variant: 'quaternary',
}

export const SmallButton = Template.bind({})
SmallButton.args = {
  ...Default.args,
  shrink: true,
}

export const ButtonWithLeftIcon = Template.bind({})
ButtonWithLeftIcon.args = {
  ...Default.args,
  leftIcon: require('@assets/icon.png'),
  variant: 'secondary',
  title: 'My Icon Button',
}

export const ButtonWithRightIcon = Template.bind({})
ButtonWithRightIcon.args = {
  ...Default.args,
  rightIcon: require('@assets/favicon.png'),
  title: 'My Icon Button',
}
