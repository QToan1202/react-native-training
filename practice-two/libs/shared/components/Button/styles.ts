import {
  ButtonIcon,
  GetProps,
  SizableText,
  Stack,
  createStyledContext,
  styled,
  withStaticProperties,
} from 'tamagui'

const ButtonContext = createStyledContext({
  color: undefined,
  ellipse: undefined,
  fontFamily: undefined,
  fontSize: undefined,
  fontStyle: undefined,
  fontWeight: undefined,
  letterSpacing: undefined,
  maxFontSizeMultiplier: undefined,
  size: undefined,
  textAlign: undefined,
  lineHeight: undefined,
})

const CustomButtonFrame = styled(Stack, {
  name: 'Button',
  context: ButtonContext,
  paddingVertical: '$space.3',
  borderRadius: '$radius.12',
  borderWidth: '$space.1',
  borderColor: '$color.transparent',
  display: 'flex',
  flexDirection: 'row',
  justifyContent: 'center',
  alignItems: 'center',
  columnGap: 6, // getTokens().space[2].val - 1
  pressStyle: {
    opacity: 0.5,
  },

  variants: {
    variant: {
      primary: {
        borderColor: '$color.primary',
        backgroundColor: '$color.primary',
        color: '$color.white',
      },

      secondary: {
        borderColor: '$color.transparent',
        backgroundColor: '$color.secondary',
        color: '$color.primary',
      },

      tertiary: {
        borderColor: '$color.white',
        backgroundColor: '$color.transparent',
        color: '$color.white',
      },

      quaternary: {
        borderRadius: '$radius.0',
        borderColor: '$color.white',
        backgroundColor: '$color.white',
        color: '$color.black',
      },
    },

    shrink: {
      true: {
        alignSelf: 'flex-start',
        paddingVertical: '$space.2',
        paddingHorizontal: 6,
      },
    },

    isDisable: {
      true: {
        opacity: 0.5,
      },
    },
  } as const,

  defaultVariants: {
    variant: 'primary',
  },
})

const ButtonText = styled(SizableText, {
  name: 'Button',
  context: ButtonContext,
  userSelect: 'none',

  variants: {
    default: {
      true: {
        fontWeight: '$2',
        textTransform: 'capitalize',
        textAlign: 'center',
        lineHeight: '$2',
      },
    },
    size: {
      '...fontSize': (name, { font }) => ({
        fontSize: font?.size[name],
      }),
    },
  } as const,

  defaultVariants: {
    default: true,
  },
})

const ButtonStyled = withStaticProperties(CustomButtonFrame, {
  Text: ButtonText,
  Icon: ButtonIcon,
  Props: ButtonContext.Provider,
})

export type ButtonProps = GetProps<typeof ButtonText> & GetProps<typeof CustomButtonFrame>

export default ButtonStyled
