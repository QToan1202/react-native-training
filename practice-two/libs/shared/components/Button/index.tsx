import { memo, useMemo } from 'react'
import { Image, ImageSourcePropType } from 'react-native'
import isEqual from 'react-fast-compare'

import { imageStyles } from '../../styles'

import ButtonStyled, { ButtonProps as StyledButtonProps } from './styles'

export type ButtonProps = {
  title: string
  leftIcon?: ImageSourcePropType
  rightIcon?: ImageSourcePropType
} & StyledButtonProps

const Button = ({ title, leftIcon, rightIcon, isDisable, onPress, ...rest }: ButtonProps) => {
  const onPressWithDisableState = useMemo(
    () => (isDisable ? undefined : onPress),
    [isDisable, onPress]
  )

  return (
    <ButtonStyled {...rest} isDisable={isDisable} onPress={onPressWithDisableState}>
      {leftIcon && (
        <ButtonStyled.Icon>
          <Image source={leftIcon} style={imageStyles.icon} />
        </ButtonStyled.Icon>
      )}
      <ButtonStyled.Text>{title}</ButtonStyled.Text>
      {rightIcon && (
        <ButtonStyled.Icon>
          <Image source={rightIcon} style={imageStyles.icon} />
        </ButtonStyled.Icon>
      )}
    </ButtonStyled>
  )
}

export default memo(Button, isEqual)
