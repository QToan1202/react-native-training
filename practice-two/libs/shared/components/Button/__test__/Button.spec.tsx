import { render, fireEvent, screen } from '@testing-library/react-native'
import Button from '../index'

const mockOnPress = jest.fn()

describe('Testing button component', () => {
  beforeEach(() => {
    render(
      <>
        <Button title="primary button" variant="primary" onPress={mockOnPress} />
        <Button title="secondary button" variant="secondary" onPress={mockOnPress} />
        <Button title="quaternary button" variant="quaternary" onPress={mockOnPress} />
        <Button title="tertiary button" variant="tertiary" onPress={mockOnPress} />
        <Button title="default button" onPress={mockOnPress} />
      </>
    )
  })

  it('Render button component with primary variant', () => {
    const component = render(
      <Button
        title="press here"
        variant="primary"
        leftIcon={require('@assets/icon.png')}
        rightIcon={require('@assets/icon.png')}
        onPress={mockOnPress}
      />
    ).toJSON()

    expect(component).toMatchSnapshot()
  })

  it('Press on button with fire onPress event', () => {
    const btn = screen.getByText('default button')

    fireEvent.press(btn)
    expect(mockOnPress).toBeCalledTimes(1)
  })
})
