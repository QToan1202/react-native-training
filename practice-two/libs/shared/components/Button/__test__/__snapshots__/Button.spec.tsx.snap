// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing button component Render button component with primary variant 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "backgroundColor": "#33907C",
      "borderColor": "#33907C",
      "borderRadius": 50,
      "borderWidth": 2,
      "opacity": 1,
      "paddingVertical": 13,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "columnGap": 6,
        "display": "flex",
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <Image
      source={
        {
          "testUri": "../../../assets/icon.png",
        }
      }
      style={
        {
          "height": 24,
          "width": 24,
        }
      }
    />
    <Text
      style={
        [
          {
            "color": "#fff",
            "fontFamily": "Montserrat_400Regular",
          },
          {
            "fontSize": 16,
          },
          [
            {
              "color": "#fff",
            },
            {
              "fontFamily": "Montserrat_500Medium",
              "textAlign": "center",
              "textTransform": "capitalize",
            },
            undefined,
          ],
        ]
      }
    >
      press here
    </Text>
    <Image
      source={
        {
          "testUri": "../../../assets/icon.png",
        }
      }
      style={
        {
          "height": 24,
          "width": 24,
        }
      }
    />
  </View>
</View>
`;
